src/main/scala/cromwell/binding/Scope.scala

```diff
@@ -1,14 +1,168 @@
 package cromwell.binding
 
+import cromwell.binding.AstTools.AstNodeName
+import cromwell.parser.WdlParser.{AstList, Ast, AstNode}
+import scala.collection.JavaConverters._
+
+
+object Scope {
+
+  /**
+   * Utility class to generate a ScopeTree from an Ast.
+   */
+  private class ScopeTreeGenerator {
+
+    /* Keep track of the # of created Scatter nodes. */
+    var scatterCount: Int = 0
```

Comment:
Ok so I wrote 2 new versions. Both versions use a new trait called `IndexedScope` which only extends `Scope` and add a `val index: Int` attribute.

- First version is still mutable, without var but with a mutable map that handles indexes for all class extending `IndexedScope`. This will avoid the multiplication of parameters/attribute when we'll add new scopes.
- Second version is completely immutable. No var and an immutable map with the same mechanism to handle `IndexedScope`s. 

IMO the first one is a lot easier to understand but immutability wins i guess :p. Here are both versions if you'd like to have a look before I commit : 
version with mutable map : https://gist.github.com/Horneth/a152825f35221cdc1ff3
version fully immutable : https://gist.github.com/Horneth/59465edf859acae259f6